File: iTunda.Web/src/lib/currency.ts
// Multi-currency support. All prices in the app are stored/quoted in KES;
// this module converts + formats them into the visitor's chosen currency.
// Rates are fetched live (base = KES) with a static fallback so the app always
// works even if the FX endpoint is unavailable.

export interface CurrencyDef {
  code: string;
  symbol: string;
  name: string;
  countryCode: string; // for flagcdn
}

export const CURRENCIES: CurrencyDef[] = [
  { code: 'KES', symbol: 'KSh', name: 'Kenyan Shilling', countryCode: 'ke' },
  { code: 'USD', symbol: '$', name: 'US Dollar', countryCode: 'us' },
  { code: 'EUR', symbol: '€', name: 'Euro', countryCode: 'eu' },
  { code: 'GBP', symbol: '£', name: 'British Pound', countryCode: 'gb' },
  { code: 'ZAR', symbol: 'R', name: 'South African Rand', countryCode: 'za' },
  { code: 'AED', symbol: 'د.إ', name: 'UAE Dirham', countryCode: 'ae' },
  { code: 'INR', symbol: '₹', name: 'Indian Rupee', countryCode: 'in' },
  { code: 'CNY', symbol: '¥', name: 'Chinese Yuan', countryCode: 'cn' },
  { code: 'NGN', symbol: '₦', name: 'Nigerian Naira', countryCode: 'ng' },
  { code: 'BRL', symbol: 'R$', name: 'Brazilian Real', countryCode: 'br' },
];

// Value of 1 KES in each currency (static fallback, refreshed periodically).
export const FALLBACK_RATES: Record<string, number> = {
  KES: 1, USD: 0.0077, EUR: 0.0071, GBP: 0.0061, ZAR: 0.14,
  AED: 0.028, INR: 0.64, CNY: 0.056, NGN: 11.8, BRL: 0.042,
};

export function currencyDef(code: string): CurrencyDef {
  return CURRENCIES.find(c => c.code === code) ?? CURRENCIES[0];
}

const RATES_KEY = 'itunda_fx';
const RATES_TTL = 6 * 60 * 60 * 1000; // 6 hours

export async function fetchRates(): Promise<Record<string, number>> {
  try {
    const cached = localStorage.getItem(RATES_KEY);
    if (cached) {
      const { at, rates } = JSON.parse(cached);
      if (Date.now() - at < RATES_TTL && rates?.USD) return { ...FALLBACK_RATES, ...rates };
    }
  } catch { /* ignore */ }

  try {
    const res = await fetch('https://open.er-api.com/v6/latest/KES', { signal: AbortSignal.timeout(5000) });
    const d = await res.json();
    if (d?.rates?.USD) {
      const rates: Record<string, number> = {};
      for (const c of CURRENCIES) if (d.rates[c.code] != null) rates[c.code] = d.rates[c.code];
      rates.KES = 1;
      localStorage.setItem(RATES_KEY, JSON.stringify({ at: Date.now(), rates }));
      return { ...FALLBACK_RATES, ...rates };
    }
  } catch { /* fall through */ }

  return { ...FALLBACK_RATES };
}

export function convertFromKes(amountKes: number, code: string, rates: Record<string, number>): number {
  return amountKes * (rates[code] ?? FALLBACK_RATES[code] ?? 1);
}

export function formatMoney(
  amountKes: number,
  code: string,
  rates: Record<string, number>,
  opts: { withCode?: boolean } = {},
): string {
  const def = currencyDef(code);
  const value = convertFromKes(amountKes, code, rates);
  const decimals = Math.abs(value) >= 100 || value === 0 ? 0 : 2;
  const num = value.toLocaleString('en-US', { minimumFractionDigits: decimals, maximumFractionDigits: decimals });
  const symbolFirst = !['د.إ'].includes(def.symbol);
  const body = symbolFirst ? `${def.symbol}${num}` : `${num} ${def.symbol}`;
  return opts.withCode ? `${body} ${def.code}` : body;
}

File: iTunda.App/Services/Currency.cs
using System.Net.Http.Json;
using System.Text.Json;

namespace iTunda.App.Services;

// App-wide currency selection + conversion. Prices are stored in KES; this
// converts/formats them into the user's chosen currency. Live rates are pulled
// from open.er-api.com with sensible static fallbacks so it always works offline.
public static class Currency
{
    public record Info(string Code, string Symbol, int Decimals);

    public static readonly Info[] All =
    {
        new("KES", "KSh ", 0),
        new("USD", "$", 2),
        new("EUR", "\u20AC", 2),
        new("GBP", "\u00A3", 2),
        new("AED", "AED ", 2),
        new("ZAR", "R", 2),
        new("NGN", "\u20A6", 0),
        new("INR", "\u20B9", 0),
    };

    public static Info Current { get; private set; } = All[0];

    // Target units per 1 KES (static fallback; refreshed by LoadRatesAsync).
    static readonly Dictionary<string, decimal> Rates = new()
    {
        ["KES"] = 1m,
        ["USD"] = 0.0077m,
        ["EUR"] = 0.0071m,
        ["GBP"] = 0.0061m,
        ["AED"] = 0.028m,
        ["ZAR"] = 0.14m,
        ["NGN"] = 11.5m,
        ["INR"] = 0.65m,
    };

    public static event Action? Changed;

    public static void Set(string code)
    {
        var info = All.FirstOrDefault(x => x.Code == code);
        if (info is not null && info.Code != Current.Code)
        {
            Current = info;
            Changed?.Invoke();
        }
    }

    public static decimal Convert(decimal kes) =>
        kes * (Rates.TryGetValue(Current.Code, out var r) ? r : 1m);

    public static string Format(decimal kes)
    {
        var v = Convert(kes);
        return Current.Decimals == 0 ? $"{Current.Symbol}{v:N0}" : $"{Current.Symbol}{v:N2}";
    }

    static bool _loaded;

    public static async Task LoadRatesAsync()
    {
        if (_loaded) return;
        try
        {
            using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(8) };
            var doc = await http.GetFromJsonAsync<JsonElement>("https://open.er-api.com/v6/latest/KES");
            if (doc.TryGetProperty("rates", out var rates) && rates.ValueKind == JsonValueKind.Object)
            {
                foreach (var info in All)
                {
                    if (rates.TryGetProperty(info.Code, out var v) && v.TryGetDecimal(out var d) && d > 0)
                        Rates[info.Code] = d;
                }
                _loaded = true;
            }
        }
        catch
        {
            // keep static fallbacks
        }
    }
}
